src/Test.EndToEnd.Baselining/Extensions/ArtifactUriEnumeration.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Sarif;
+
+namespace SarifBaseline.Extensions
+{
+    public static class ArtifactUriEnumeration
+    {
+        public static IEnumerable<Uri> AllArtifactUris(this SarifLog log)
+        {
+            if (log != null)
+            {
+                foreach (Run run in log.EnumerateRuns())
```

Comment:
>EnumerateRuns [](start = 40, length = 13)

I'm not convinced of the value adding an API to avoid a single null check makes sense. Better would be for the emitted SARIF code to have an accessor that auto-populated log.Runs to a non-null but empty collection for this case. We would want, naturally, deserialization code that elides non-null but empty things (we have this elsewhere).

Reply:
We shouldn't auto-populate `log.runs` because `null` is semantically different from empty for this property. See §3.13.4.

Reply:
Basically, I constantly ended up writing:

if(log?.Runs != null)
{
      foreach(Run run in log.Runs)
      {
            if(run?.Results != null)
            {
                  foreach(Result result in run.Results)
                  {
                         if(result?.Locations != null)
                         {
                                  ...
}}}}}

Basically, each one of these methods merges a null check and foreach loop into one thing. I can then build each extension method on the one above (for EnumerateResults(SarifLog)) and a user can start out anywhere in the SarifLog hierarchy without double statements everywhere.

I agree this method doesn't do a lot, but I think it's useful in making enumeration code simpler and consistent for enumerating from any level to any level in the Sarif hierarchy. That's the overall point.